tests/test_entrypoint.py

```diff
@@ -0,0 +1,239 @@
+"""
+Testing the logic that extracts the entrypoint data:
+    - CLI entrypoints
+    - GUI entrypoints
+    - miscellaneous entrypoints
+"""
+
+from setuptools_pyproject_migration import WritePyproject
+from setuptools.dist import Distribution
+
+
+# Entry point parsing
+
+
+def test_parse_entrypoint_happypath():
+    """
+    Test we can extract the entry point name and target from a string.
+    """
+    assert WritePyproject._parse_entrypoint("ep=project.module:target") == ("ep", "project.module:target")
+
+
+def test_parse_entrypoint_happypath_whitespace():
+    """
+    Test we strip extraneous whitespace from the strings received.
+    """
+    assert WritePyproject._parse_entrypoint("  ep  =  project.module:target  ") == ("ep", "project.module:target")
+
+
+def test_parse_entrypoint_missing_eq():
+    """
+    Test that a string lacking a '=' is refused.
+    """
+    try:
+        WritePyproject._parse_entrypoint("ep project.module:target")
+        assert False, "Should not have been accepted as valid"
+    except ValueError as e:
+        assert str(e).endswith(" is not of the form 'name = module:function'")
+
+
+# Wrapper function logic
+
+
+def test_generate_entrypoints_none():
+    """
+    Test we get an empty dict if entrypoints is None.
+    """
+    assert WritePyproject._generate_entrypoints(None) == {}
+
+
+def test_generate_entrypoints():
+    """
+    Test we get all entry points grouped by type.
+    """
+    assert WritePyproject._generate_entrypoints(
+        {
+            "console_scripts": [
+                "spanish-inquisition = montypython.unexpected:spanishinquisition",
+                "brian=montypython.naughtyboy:brian",
+            ],
+            "gui_scripts": [
+                "dead-parrot=montypython.sketch:petshop",
+                "shrubbery=montypython.holygrail:knightswhosayni",
+            ],
+            "project.plugins": [
+                "babysnatchers=montypython.somethingcompletelydifferent:babysnatchers",
+                "eels=montypython.somethingcompletelydifferent:eels",
+            ],
+        }
+    ) == {
+        "console_scripts": {
+            "spanish-inquisition": "montypython.unexpected:spanishinquisition",
+            "brian": "montypython.naughtyboy:brian",
+        },
+        "gui_scripts": {
+            "dead-parrot": "montypython.sketch:petshop",
+            "shrubbery": "montypython.holygrail:knightswhosayni",
+        },
+        "project.plugins": {
+            "babysnatchers": "montypython.somethingcompletelydifferent:babysnatchers",
+            "eels": "montypython.somethingcompletelydifferent:eels",
+        },
+    }
+
+
+# Test correct placement of the parsed endpoints
```

Comment:
I noticed that the rest of the tests in this file are, I guess, integration tests (i.e. testing the overall operation of the plugin rather than testing individual functions like `_generate_entrypoints()`), and it got me thinking that personally I would probably split those into a separate file to make the distinction between the two kinds of tests more clear, and also to make it easier to run only one or the other kind. (Integration tests take longer, if nothing else) But honestly that's just my personal preference and I don't know if it's really the best way to do it... Anyway, I am fine to go forward either way, but it's something to think about before finalizing this PR.

It is definitely a great thing to _have_ tests that just feed a manually-constructed `Distribution` object to `WritePyproject` though. I should have written more of those in my previous work, and I definitely want to follow your lead and lean more heavily on them in the future.

Reply:
Yeah, I sort of grouped these all in since I'm specifically looking at the entry point outputs (and ignoring everything else).

I think the big distinction is that most of the tests of this nature are invoking the TOML serialiser, whereas this is comparing the parts of the `dict` directly.  Those are definitely "integration" tests since about the only thing we _don't_ do is generate a dummy `setup.py` and run it in a sub-process, we're not far removed from that though.

As for time taken:

```
======================================================= test session starts =======================================================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0 -- /home/stuartl/projects/setuptools-pyproject-migration/env/bin/python3
cachedir: .pytest_cache
rootdir: /home/stuartl/projects/setuptools-pyproject-migration/work
configfile: pytest.ini
plugins: cov-4.1.0, console-scripts-1.4.1
collected 10 items                                                                                                                

tests/test_entrypoint.py::test_parse_entrypoint_happypath PASSED                                                            [ 10%]
tests/test_entrypoint.py::test_parse_entrypoint_happypath_whitespace PASSED                                                 [ 20%]
tests/test_entrypoint.py::test_parse_entrypoint_missing_eq PASSED                                                           [ 30%]
tests/test_entrypoint.py::test_generate_entrypoints_none PASSED                                                             [ 40%]
tests/test_entrypoint.py::test_generate_entrypoints PASSED                                                                  [ 50%]
tests/test_entrypoint.py::test_generate_noentrypoints PASSED                                                                [ 60%]
tests/test_entrypoint.py::test_generate_clionly PASSED                                                                      [ 70%]
tests/test_entrypoint.py::test_generate_guionly PASSED                                                                      [ 80%]
tests/test_entrypoint.py::test_generate_misconly PASSED                                                                     [ 90%]
tests/test_entrypoint.py::test_generate_all_entrypoints PASSED                                                              [100%]

======================================================= 10 passed in 0.12s ========================================================
```

This machine is definitely no spring chicken:
```
RC=0 stuartl@rikishi ~/projects/setuptools-pyproject-migration/work $ grep '^model name' /proc/cpuinfo | head -n 1
model name      : Intel(R) Core(TM) i5-3320M CPU @ 2.60GHz
```

I do note the `test_everything` test does take a lot longer (almost 25 seconds), not quite sure why at this point, but I'm not overly bothered either… 30 seconds is not long to wait for test results.

Reply:
Sure, fair enough.

Personally I find 25 seconds long enough to break concentration, so I do tend to avoid that test when I'm quickly iterating on things. But yeah, it's not the biggest deal. Perhaps it takes a while because it's installing dependencies.